torch/csrc/jit/codegen/cuda/expr_simplifier.cpp

```diff
@@ -833,14 +941,83 @@ Val* eliminateTrivialComputation(Val* value) {
+      if (bop->lhs()->sameAs(bop->rhs())) {
+        if (op == BinaryOpType::Eq) {
+          return IrBuilder::newConstant(true, DataType::Bool);
+        } else if (op == BinaryOpType::NE) {
+          return IrBuilder::newConstant(false, DataType::Bool);
+        }
+      }
+      if (bop->rhs()->isZeroInt()) {
+        if (op == BinaryOpType::GE &&
+            prove::isNonNegative(bop->lhs(), var_info)) {
+          return IrBuilder::newConstant(true, DataType::Bool);
+        } else if (
+            op == BinaryOpType::GT && prove::isPositive(bop->lhs(), var_info)) {
+          return IrBuilder::newConstant(true, DataType::Bool);
+        } else if (
+            op == BinaryOpType::NE && prove::isNonZero(bop->lhs(), var_info)) {
+          return IrBuilder::newConstant(true, DataType::Bool);
+        } else if (
+            op == BinaryOpType::LT &&
+            prove::isNonNegative(bop->lhs(), var_info)) {
+          return IrBuilder::newConstant(false, DataType::Bool);
+        } else if (
+            op == BinaryOpType::LE && prove::isPositive(bop->lhs(), var_info)) {
+          return IrBuilder::newConstant(false, DataType::Bool);
+        } else if (
+            op == BinaryOpType::Eq && prove::isNonZero(bop->lhs(), var_info)) {
+          return IrBuilder::newConstant(false, DataType::Bool);
+        }
+      } else if (bop->lhs()->isZeroInt()) {
+        if (op == BinaryOpType::LE &&
+            prove::isNonNegative(bop->rhs(), var_info)) {
+          return IrBuilder::newConstant(true, DataType::Bool);
+        } else if (
+            op == BinaryOpType::LT && prove::isPositive(bop->rhs(), var_info)) {
+          return IrBuilder::newConstant(true, DataType::Bool);
+        } else if (
+            op == BinaryOpType::NE && prove::isNonZero(bop->rhs(), var_info)) {
+          return IrBuilder::newConstant(true, DataType::Bool);
+        } else if (
+            op == BinaryOpType::GT &&
+            prove::isNonNegative(bop->rhs(), var_info)) {
+          return IrBuilder::newConstant(false, DataType::Bool);
+        } else if (
+            op == BinaryOpType::GE && prove::isPositive(bop->rhs(), var_info)) {
+          return IrBuilder::newConstant(false, DataType::Bool);
+        } else if (
+            op == BinaryOpType::Eq && prove::isNonZero(bop->rhs(), var_info)) {
+          return IrBuilder::newConstant(false, DataType::Bool);
+        }
+      }
+    }
+  }
+  return value;
+}
+
 } // namespace rules
 
-#define RUN_PASS(pass_name)                               \
-  simplified = recurseDown(simplified, rules::pass_name); \
+#define RUN_PASS(pass_name)                                    \
+  simplified = recurseDown(simplified, [&var_info](Val* val) { \
+    return rules::pass_name(val, var_info);                    \
+  });                                                          \
   logger->record(#pass_name, simplified)
 
 Val* simplifyExpr(Val* value, const std::list<ValInfo>& variables) {
```

Comment:
Trying to remember how this is used. In `ForLoop::simplifiedStop()`, no variable is given. Is it fine because no stop expr should use loop indices?

Reply:
Yes, exactly. The second argument, `variables`, is optional additional information. This additional information makes `simplifyExpr` possible to do more simplification, but no matter whether the additional information is provided or not, `simplifyExpr` always does correct simplifications. So, it is always OK not to provide such information.